### PR TITLE
Full enablement of Unions using nightly Rust

### PR DESCRIPTION
**Pull request by Luke-Nukem**
Unions are fully enabled behind a feature-gate. Both non-nightly and nightly Rust appear to be working well.

### CODE REVIEW

**Comment by EPashkin** [comment]
@Luke-Nukem Can you rebase this to master with removing merges?
Why `GVariantBuilder_u1` etc. added to all generated libs in https://github.com/gtk-rs/sys when this feature enabled?
Bad changes in glib-sys with bitsized fields: GDate as example
```diff
#[repr(C)]
 pub struct GDate {
-    _truncated_record_marker: c_void,
-    //julian_days: guint: 32,
-    //julian: guint: 1,
-    //dmy: guint: 1,
-    //day: guint: 6,
-    //month: guint: 4,
-    //year: guint: 16,
+    pub julian_days: c_uint,
+    pub julian: c_uint,
+    pub dmy: c_uint,
+    pub day: c_uint,
+    pub month: c_uint,
+    pub year: c_uint,
 }
```

Strange that GScanner has `//union` comments before, but now it seems fixed, thanks

File: src/parser.rs [hunk 1]
```diff
@@ -362,9 +438,10 @@ impl Library {
             }
         }
 
-        if attrs.by_name("is-gtype-struct").is_some() {
-            return Ok(());
-        }
+        // TODO: unsure of how to handle this
+        //if attrs.by_name("is-gtype-struct").is_some() {
+        //    return Ok(());
+        //}
```
Comment (EPashkin) [inline src/parser.rs]: Seems `read_record` return type need change to `Result<Option<Type>>`

**Comment by Luke-Nukem** [comment]
Regarding the appending of `_u1` to various unions or `_s1` to nested union->structs, it's because without it we end up with each struct->union->struct having the same c_type, and thus the same name for each.

Regarding
```
#[repr(C)]
 pub struct GDate {
-    _truncated_record_marker: c_void,
-    //julian_days: guint: 32,
-    //julian: guint: 1,
-    //dmy: guint: 1,
-    //day: guint: 6,
-    //month: guint: 4,
-    //year: guint: 16,
+    pub julian_days: c_uint,
+    pub julian: c_uint,
+    pub dmy: c_uint,
+    pub day: c_uint,
+    pub month: c_uint,
+    pub year: c_uint,
 }
```
Well of course... We can't initialise a newly created struct like that. Have to declare it first, then initialise it. If it were an enum instead of a struct then it might work.

If you look at;
```
struct GDate {
  guint julian_days : 32; /* julian days representation - we use a
                           *  bitfield hoping that 64 bit platforms
                           *  will pack this whole struct in one big
                           *  int
                           */

  guint julian : 1;    /* julian is valid */
  guint dmy    : 1;    /* dmy is valid */

  /* DMY representation */
  guint day    : 6;
  guint month  : 4;
  guint year   : 16;
};
```
each field is a guint, which translates to a c_uint. And they are initialised (which we can't do in rust).

File: src/parser.rs (hunk 1 above)
Comment (Luke-Nukem) [inline src/parser.rs]: Yes but it also needs an initialised Type to return. This is where I am unsure what to do... I could maybe return a `default()` initialised Type?

**Comment by EPashkin** [comment]
As I remember size GDate in C will be 64bit or 8bytes with `julian`, `dmy` and `day` stored in byte 5)
Your version: 6*4=24 bytes;

File: src/parser.rs (hunk 1 above)
Comment (EPashkin) [inline src/parser.rs]: Why it can't return `Ok(None)` in this case? And don't add type in `read_record_start`

Comment (Luke-Nukem) [inline src/parser.rs]: `Result<Option<Type>>`.... I don't know why my mind completely ignored that....
Done.

**Comment by Luke-Nukem** [comment]
> As I remember size GDate in C will be 64bit or 8bytes with julian, dmy and day stored in byte 5)
> Your version: 6*4=24 bytes;

Looking at [guint](https://developer.gnome.org/glib/stable/glib-Basic-Types.html#guint) it seems it will be a particular size depending on a `#define G_MAXUINT UINT_MAX`.

 `UINT_MAX` is defined in Rust too, and c_uint size is defined by that. So essentially it looks like both `guint` and `c_uint` are defined the same way, and so should be the same size.

File: src/parser.rs [hunk 2]
```diff
@@ -21,6 +21,7 @@ pub fn is_empty_c_type(c_type: &str) -> bool {
     c_type == EMPTY_CTYPE
 }
 
+
```
Comment (EPashkin) [inline src/parser.rs]: Unneeded extra line

File: src/parser.rs [hunk 3]
```diff
@@ -322,24 +323,19 @@ impl Library {
...
-        let mut c_type = try!(
-            attrs
-                .by_name("type")
-                .ok_or_else(|| mk_error!("Missing c:type attribute", parser))
-        );
+        -> Result<Option<Type>> {
+        let mut name = try!(attrs.by_name("name")
+                .ok_or_else(|| mk_error!("Missing record name", parser)));
+        let mut c_type = try!(attrs.by_name("type")
+                .ok_or_else(|| mk_error!("Missing c:type attribute", parser)));
```
Comment (EPashkin) [inline src/parser.rs]: [rustfmt-nigthly](https://crates.io/crates/rustfmt-nightly) change this 2 variables back.

**Comment by EPashkin** [comment]
Seems you ignore that
```
guint julian : 1;
guint dmy : 1; 
```
stored in one byte and all struct size can be lesser that guint size (not *2 as in your case).
Rust IMHO don't support this syntax,
this why we "ignore" stucts with these fields.
good that seems none of these structs used in unions,
so you can fallback to old code for this case

Also still not solved problem with including `GVariantBuilder_u1` and `GVariantBuilder_s1` without next usage to all your sys crates.

**Comment by Luke-Nukem** [comment]
```
guint julian : 1;
guint dmy : 1; 
```
Hmm, I thought that was initialising those to the integer number 'n'...

>Also still not solved problem with including GVariantBuilder_u1 and GVariantBuilder_s1 without next usage to all your sys crates.

I don't understand what you mean... The result is;
```
#[repr(C)]
pub struct GVariantBuilder {
    pub u1: GVariantBuilder_u1,
}
 
#[repr(C)]
pub union GVariantBuilder_u1 {
    pub s1: GVariantBuilder_s1,
    pub x: [size_t; 16],
}
 
#[repr(C)]
pub struct GVariantBuilder_s1 {
    pub partial_magic: size_t,
    pub type_: *const GVariantType,
    pub y: [size_t; 14],
}
```
which to me seems fine. You would need to write the API on top of this to accommodate the nesting... 

I think the whole issue comes from in the gir;
```
   <record (trimmed info)>
      <union name="u" c:type="u">
        <record name="s" c:type="s">
```
I can't seem to get those names out, and if I did and the C_Type ended up being name 'u' or 's' then this would cause issues. I can make a small change to not add a number to those two fields, but then I'm not sure it that will cause issues later either.

**Comment by EPashkin** [comment]
Result is fine. Problem that it added to all sys crates not only on glib (7 unused copies for each these union and inner struct).

**Comment by EPashkin** [comment]
Same with `GDoubleIEEE754_s1`, `GFloatIEEE754_s1`.
Problem at seems that `Library::record` and `Library::union` add type to MAIN_NAMESPACE

**Comment by Luke-Nukem** [comment]
Ah I see.. Yes I was thinking about adding special cases, but then I tried to make it general.

```
    #[cfg(feature = "use_unions")]
    pub fn union(library: &mut Library, u: Union) -> TypeId {
        let fields = u.fields;
        let field_tids: Vec<TypeId> = fields.iter().map(|f| f.typ).collect();
        let typ = Type::Union(Union {
            name            : u.name,
            c_type          : u.c_type,
            fields          : fields,
            functions       : u.functions,
            doc             : u.doc,
        });
        library.add_type(MAIN_NAMESPACE, &format!("#{:?}", field_tids), typ)
    }
```
Is it that do you think? 

Regarding the bit field structs, I'll see if I can use [bitflags](https://github.com/rust-lang-nursery/bitflags) crate.

**Comment by EPashkin** [comment]
Yes, it one of functions, that cause problem (I thought that it in Library). It need `ns_id: u16` parameter.
Bitflags IMHO don't help with this and better just use "old" code with commenting.

**Comment by Luke-Nukem** [comment]
@EPashkin are you sure about the bitflags? Mozilla is using it with bindgen for their servo stuff, and the results seem to wok well.

**Comment by EPashkin** [comment]
You can try ;)

**Comment by EPashkin** [comment]
Strange, but duplicates still present :(

File: src/parser.rs [hunk 4]
```diff
@@ -302,18 +319,32 @@ impl Library {
...
+        if let Some(typ) = try!(self.read_record(parser, ns_id, attrs)) {
+            let name = typ.get_name().clone();
+            self.add_type(ns_id, &name, typ);
+        }
+        Ok(())
+    }
+
+    fn read_record(&mut self, parser: &mut Reader, ns_id: u16, attrs: &Attributes)
+        -> Result<Option<Type>> {
+        let name = try!(
```
Comment (EPashkin) [inline src/parser.rs]: let mut name

Comment (Luke-Nukem) [inline src/parser.rs]: Doh!

**Comment by Luke-Nukem** [comment]
I'm not seeing any duplicates here 😕

**Comment by EPashkin** [comment]
I meant that (for example) both atk-sys/src/lib.rs and glib-sys/src/lib.rs
have `GVariantBuilder_u1`, and it don't used in `atk-sys` at all.

**Comment by EPashkin** [comment]
Sorry, something bugged on my machine: duplicate actually removed.
Only last problem with bitsized structs

**Comment by EPashkin** [comment]
Minor bug in result formating:
https://github.com/gtk-rs/gir/pull/402/files#diff-0c37b0c87c3c91119e579feee853b5d0R368 only needed if feature disabled.

**Comment by Luke-Nukem** [comment]
Okay, that should be the final issue. Except the result formatting? I can't see what you are referring too.

**Comment by EPashkin** [comment]
Seems github generate broken links.
I meant last write in `codegen/sys/lib_::generate_unions`
```rust
    if !items.is_empty() {
        try!(writeln!(w, ""));
    }

```

**Comment by EPashkin** [comment]
Not produce strange results for GIOChannel, GScannerConfig,  gobject-sys::GClosure. gtk-sys::GtkTextAttributes.
Before all after bit-sized fields was commented too.
```rust
#[repr(C)]
pub struct GIOChannel {
    pub ref_count: c_int,
    pub funcs: *mut GIOFuncs,
    pub encoding: *mut c_char,
    pub read_cd: GIConv,
    pub write_cd: GIConv,
    pub line_term: *mut c_char,
    pub line_term_len: c_uint,
    pub buf_size: size_t,
    pub read_buf: *mut GString,
    pub encoded_read_buf: *mut GString,
    pub write_buf: *mut GString,
    pub partial_write_buf: [c_char; 6],
    _truncated_record_marker: c_void,
    //use_buffer: guint: 1,
    //do_encode: guint: 1,
    //close_on_unref: guint: 1,
    //is_readable: guint: 1,
    //is_writeable: guint: 1,
    //is_seekable: guint: 1,
    pub reserved1: gpointer,
    pub reserved2: gpointer,
}
```

**Comment by EPashkin** [comment]
Hm, Seems https://github.com/rust-lang/rust/pull/42068 already stabilized part of unions and it in already beta.
Glib has only problem with bit-sized `GDoubleIEEE754_s1` and `GFloatIEEE754_s1` as it non-Copy.

**Comment by EPashkin** [comment]
Good thing that I don't see usage `GFloatIEEE754` and `GDoubleIEEE754` in any functions and we can try ignore its after next release.

**Comment by Luke-Nukem** [comment]
We're almost there 😄 

I see why the proper truncation is required for fields after bitflags. Just needed to allow either is_bits or !truncated.

**Comment by EPashkin** [comment]
Thanks, @Luke-Nukem, Now I don't see any potential problems in generated code.
Can you remove empty line after `generate_unions` declaration?
And also please squash this PR to one commit as there too many merges in it.

File: src/parser.rs (hunk 3 above)
Comment (Luke-Nukem) [inline src/parser.rs]: Ah, my rustfmt is outdated. Are you running everything through rustfmt?

Comment (EPashkin) [inline src/parser.rs]: Not on regular basic, but sometimes.

Comment (EPashkin) [inline src/parser.rs]: Mostly because latest tend to split function declaration to multiple lines.

**Comment by Luke-Nukem** [comment]
I think I did the squash correctly? It's commit #f56da73

**Comment by EPashkin** [comment]
Yes, #f56da73 is good, but next is not (it merge again).
Please, reset your master branch to this commit, and force push it to your repo.

**Comment by Luke-Nukem** [comment]
Got it. Figured if there was an issue it would be an easy solve. Never done a squash before, very handy.

**Comment by Luke-Nukem** [comment]
Hold fire a few minutes, I'll clean up the commit log

**Comment by EPashkin** [comment]
Actually you added to squashed commit some changes from current master (like adding external_libraries.rs and using it) but git is smart and will ignore it on merge.
Or you can do `git rebase 9e3f4ccff85d1e` to be sure that it contains only your changes.

**Comment by Luke-Nukem** [comment]
Okay, I think we're good now?

**Comment by EPashkin** [comment]
Previous was good (based on 9e3f4ccff85d1e - current upstream master)

**Comment by EPashkin** [comment]
Just to `git rebase 9e3f4ccff85d1e` again

**Comment by EPashkin** [comment]
Yes, now it good.
Thanks again for unions and readme clarification.
I wait for CI just it case and merge

**Comment by EPashkin** [comment]
And excuse me for a lot of nitpicking 😉

**Comment by Luke-Nukem** [comment]
Nitpicking is good. You caught a few things I just didn't think about (or know), or had missed. And it's good practice for me 😉 

I'm looking forward to union stabilization in Rust. And when I get more time (or need it) I'll work on the bit-fields in multitype-structs.

**Pull request merged**